fault.c

```diff
@@ -157,10 +160,161 @@ static int kibosh_fault_read_delay_check(struct kibosh_fault_read_delay *fault,
+        goto error;
+    }
+    fault->code = code_obj->u.integer;
+    return fault;
+
+error:
+    if (fault) {
+        free(fault->prefix);
+        free(fault);
+    }
+    return NULL;
+}
+
+static char *kibosh_fault_unwritable_unparse(struct kibosh_fault_unwritable *fault)
+{
+    return dynprintf("{\"type\":\"%s\", "
+                    "\"prefix\":\"%s\", "
+                    "\"code\":%d}",
+                    KIBOSH_FAULT_TYPE_UNWRITABLE,
+                    fault->prefix,
+                    fault->code);
+}
+
+static int kibosh_fault_unwritable_check(struct kibosh_fault_unwritable *fault, const char *path,
+                                         const char *op)
+{
+    if (strcmp(op, "write") != 0) {
+        return 0;
+    }
+    if (strncmp(path, fault->prefix, strlen(fault->prefix)) != 0) {
+        return 0;
+    }
+    return fault->code;
+}
+
+/**
+ * kibosh_fault_read_corrupt
+ */
+static struct kibosh_fault_read_corrupt *kibosh_fault_read_corrupt_parse(json_value *obj)
+{
+    struct kibosh_fault_read_corrupt *fault = NULL;
+    json_value *mode_obj = NULL, *prefix_obj = NULL, *fraction_obj = NULL, *file_type_obj = NULL;
+
+    mode_obj = get_child(obj, "mode");
+    if ((!mode_obj) || (mode_obj->type != json_integer)) {
+        INFO("kibosh_fault_read_corrupt_parse: No valid \"mode\" field found in fault object.\n");
+        goto error;
+    }
+    prefix_obj = get_child(obj, "prefix");
+    if ((!prefix_obj) || (prefix_obj->type != json_string)) {
+        INFO("kibosh_fault_read_corrupt_parse: No valid \"prefix\" field found in fault object.\n");
+        goto error;
+    }
+    fault = calloc(1, sizeof(*fault));
+    if (!fault) {
+        INFO("kibosh_fault_read_corrupt_parse: OOM\n");
+        return NULL;
+    }
+    snprintf(fault->base.type, KIBOSH_FAULT_TYPE_STR_LEN, "%s", KIBOSH_FAULT_TYPE_READ_CORRUPT);
+    fault->prefix = strdup(prefix_obj->u.string.ptr);
+    if (!fault->prefix) {
+        INFO("kibosh_fault_read_corrupt_parse: OOM\n");
+        goto error;
+    }
+    file_type_obj = get_child(obj, "file_type");
+    if ((!file_type_obj) || (file_type_obj->type != json_string)) {
+        INFO("kibosh_fault_read_corrupt_parse: No valid \"file_type\" field found in fault object, will apply read_corrupt to all files.\n");
+        fault->file_type = NULL;
+    } else {
+        fault->file_type = strdup(file_type_obj->u.string.ptr);
+        if (!fault->file_type) {
+            INFO("kibosh_fault_read_corrupt_parse: OOM\n");
+            goto error;
+        }
+    }
+    fraction_obj = get_child(obj, "fraction");
+    if ((!fraction_obj) || (fraction_obj->type != json_double)) {
+        INFO("kibosh_fault_read_corrupt_parse: No valid \"fraction\" field found in fault object, will apply read_corrupt to all bits\n");
+        fault->fraction = 1.0;
+    } else {
+        fault->fraction = fraction_obj->u.dbl;
+    }
+    fault->mode = mode_obj->u.integer;
+    return fault;
+
+error:
```

Comment:
If this is common code, we can create a function to handle the free logic, passing the pointer to fault struct.

Reply:
Good suggestion! Do you think we should do this for all fault_parse functions? @cmccabe